src/cpp/core/system/PosixSystem.cpp

```diff
@@ -2425,6 +2448,42 @@ Error userBelongsToGroup(const User& user,
 }
 
 
+// first try to get the user's groups via getgrouplist. we prefer this to
+// fetching information about the group to test membership because
+// this can fail in the case where SSSD group member data is disabled
+// via ignore_group_members (see https://jhrozek.fedorapeople.org/sssd/1.9.91/man/sssd.conf.5.html)
+//
+// we fallback to the old implementation if this method did not yield a positive result
+// to ensure backwards compatibility in existing customer environments. however,
+// group member enumeration of large groups is expensive and this has been known
+// to be slow in some environments.
+//
+// if the first try fails via getgrouplist, we hold the error and attempt
+// the old implementation before returning. if both calls fail, we return the first error
+Error userBelongsToGroup(const User& user,
+                         const std::string& groupName,
+                         bool* pBelongs)
+{
+   *pBelongs = false; // default to not found
+
+   // first we try to use the getgrouplist(3) implementation to determine
+   // if the user is a member of the target groupName.
+   Error groupListError = userBelongsToGroupViaGroupList(user, groupName, pBelongs);
+
+   // if the user belongs, then we can return early.
+   // otherwise, we fallback to the legacy implementation
```

Comment:
Minor nit here on the comment: we are returning only if there was no error, not if necessarily if the user belongs to the group. And I think that's the right approach, just need to update the comment.